runtime/server/downloads.go

```diff
@@ -23,67 +22,68 @@ import (
 	"google.golang.org/protobuf/proto"
 )
 
-func BakeQuery(qry *runtimev1.Query) (string, error) {
-	if qry == nil {
-		return "", errors.New("cannot bake nil query")
+func (s *Server) Export(ctx context.Context, req *runtimev1.ExportRequest) (*runtimev1.ExportResponse, error) {
+	if !auth.GetClaims(ctx).CanInstance(req.InstanceId, auth.ReadMetrics) {
+		return nil, ErrForbidden
 	}
 
-	data, err := proto.Marshal(qry)
+	tkn, err := s.generateDownloadToken(req, auth.GetClaims(ctx).SecurityClaims())
 	if err != nil {
-		return "", err
+		return nil, status.Errorf(codes.Internal, "failed to generate download token: %s", err.Error())
 	}
 
-	data, err = gzipCompress(data)
-	if err != nil {
-		return "", err
-	}
+	out := fmt.Sprintf("/v1/download?token=%s", tkn)
 
-	return base64.URLEncoding.EncodeToString(data), nil
+	return &runtimev1.ExportResponse{
+		DownloadUrlPath: out,
+	}, nil
 }
 
-func UnbakeQuery(bakedQry string) (*runtimev1.Query, error) {
-	data, err := base64.URLEncoding.DecodeString(bakedQry)
+func (s *Server) ExportReport(ctx context.Context, req *runtimev1.ExportReportRequest) (*runtimev1.ExportReportResponse, error) {
+	c, err := s.runtime.Controller(ctx, req.InstanceId)
 	if err != nil {
-		return nil, err
+		return nil, status.Errorf(codes.Internal, "failed to get controller: %s", err.Error())
 	}
 
-	uncompressed, err := gzipDecompress(data)
+	res, err := c.Get(ctx, &runtimev1.ResourceName{Kind: runtime.ResourceKindReport, Name: req.Report}, false)
 	if err != nil {
-		// NOTE (2023-11-29): Backwards compatibility for when we didn't gzip baked queries. We can remove this in a few months.
-		uncompressed = data
+		return nil, status.Errorf(codes.Internal, "failed to get report: %s", err.Error())
 	}
 
-	qry := &runtimev1.Query{}
-	if err := proto.Unmarshal(uncompressed, qry); err != nil {
-		return nil, err
+	r, access, err := s.applySecurityPolicy(ctx, req.InstanceId, res)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	if !access {
+		return nil, status.Error(codes.NotFound, "resource not found")
 	}
 
-	return qry, nil
-}
-
-func (s *Server) Export(ctx context.Context, req *runtimev1.ExportRequest) (*runtimev1.ExportResponse, error) {
-	if !auth.GetClaims(ctx).CanInstance(req.InstanceId, auth.ReadMetrics) {
-		return nil, ErrForbidden
+	if r.GetReport() == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "resource is not a report")
 	}
 
-	if req.BakedQuery != "" {
-		qry, err := UnbakeQuery(req.BakedQuery)
-		if err != nil {
-			return nil, status.Errorf(codes.InvalidArgument, "failed to parse baked query: %s", err.Error())
-		}
+	rep := r.GetReport()
+	t := req.ExecutionTime.AsTime()
 
-		req.Query = qry
-		req.BakedQuery = ""
+	qry, err := queries.ProtoFromJSON(rep.Spec.QueryName, rep.Spec.QueryArgsJson, &t)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build export request: %w", err)
 	}
```

Comment:
Do we want to keep for one more version for backwards compatibility or small blip is okay during upgrade. I assume the upgrade order is admin, runtime and then UI ?

Reply:
I think a small blip is alright as long as it doesn't cause a crash (so no reuse of proto numbers for example)